File: Haskell/25-maximal-sum-square.hs
import Control.Monad (replicateM)
import Data.Array (Array, listArray, range, (!))


convertToIntList :: String -> [Int]
convertToIntList = map read . words


readMatrix :: Int -> IO (Array (Int, Int) Int)
readMatrix n = do
    rows <- replicateM n $ fmap convertToIntList getLine
    return $ listArray ((0, 0), (n - 1, n - 1)) [x | row <- rows, x <- row]


maximalSumSquareMatrix :: Array (Int, Int) Int -> Int -> Int -> Int
maximalSumSquareMatrix matrix n k =
    let fillPrefixSum :: Int -> Int -> Int
        fillPrefixSum 0 _ = 0
        fillPrefixSum _ 0 = 0
        fillPrefixSum i j = prefixSum ! (i-1, j) + prefixSum ! (i, j-1) - prefixSum ! (i-1, j-1) + matrix ! (i-1, j-1)
        bounds = ((0, 0), (n, n))
        prefixSum = listArray bounds [fillPrefixSum i j | (i, j) <- range bounds]
    in foldl (\acc i ->
        foldl (\inner j ->
            let ps = prefixSum ! (i+k, j+k) - prefixSum ! (i+k, j) - prefixSum ! (i, j+k) + prefixSum ! (i, j)
            in max inner ps) acc [0..(n-k)] ) 0 [0..(n-k)]


main :: IO()
main = do
    [n, k] <- fmap convertToIntList getLine
    matrix <- readMatrix n
    let result = maximalSumSquareMatrix matrix n k
    print result

File: Python/25_maximal_sum_square.py
import sys


def convert_to_intlist(line):
    return [int(item) for item in line.split()]


def read_matrix(reader, n):
    matrix = []
    for _ in range(n):
        row = convert_to_intlist(next(reader))
        matrix.append(row)
    return matrix


def maximal_sum_square_submatrix(matrix, n, k):
    prefix_sum = [[0] * (n+1) for _ in range(n+1)]
    for ix in range(1, n+1):
        for jy in range(1, n+1):
            ps = (prefix_sum[ix-1][jy] + prefix_sum[ix][jy-1] - prefix_sum[ix-1][jy-1] + matrix[ix-1][jy-1])
            prefix_sum[ix][jy] = ps

    result = 0
    for ix in range(n-k+1):
        for jy in range(n-k+1):
            s = prefix_sum[ix+k][jy+k] - prefix_sum[ix+k][jy] - prefix_sum[ix][jy+k] + prefix_sum[ix][jy]
            if s > result:
                result = s
    return result


def main():
    reader = sys.stdin
    n, k = convert_to_intlist(next(reader))
    matrix = read_matrix(reader, n)
    result = maximal_sum_square_submatrix(matrix, n, k)
    print(result)


if __name__ == "__main__":
    main()
